### PR TITLE
Text transformation (Please review this one LAST)

### PR DESCRIPTION
**Pull request by JingJZ160**
### Description

- All text columns will be concatenated together into one column. It just make no sense to me if we run TFIDF on each text column and then concatenate the output together.
- At the end of the text transformation pipeline, I added a TruncatedSVD step. PCA does not accept sparse input and as the documentation suggested, use the TruncatedSVD transformer instead.
- Current number of components is set to the same number of non-text features.
- HTMLRemover step is temporarily turned off as it makes things complicated when there are pure text columns and raw html columns. I would rather just deal with pure text for now unless we get REAL demands for processing raw html data.

### CODE REVIEW

File: foreshadow/steps/preprocessor.py [hunk 1]
```diff
@@ -43,13 +49,32 @@ def fit(self, X, *args, **kwargs):
...
+                transformation_pipeline = self.pipeline_by_intent[intent]
+                list_of_tuples.append(
+                    (column, transformation_pipeline, column)
+                )
 
+        text_transformation_pipeline = _configure_text_transformation_pipeline(
+            num_of_non_text_features=len(X.columns) - len(text_features)
+        )
+        list_of_tuples.append(
+            ("text_preprocessing", text_transformation_pipeline, text_features)
```
Comment (jichaogp) [inline foreshadow/steps/preprocessor.py]: Where does the merge of all text_features happen?

File: foreshadow/steps/preprocessor.py [hunk 2]
```diff
@@ -88,6 +113,17 @@ def fit_transform(self, X, *args, **kwargs):
...
+            and override_key in self.cache_manager[AcceptedKey.OVERRIDE]
+        ):
+            intent = self.cache_manager[AcceptedKey.OVERRIDE][override_key]
+        else:
+            intent = self.cache_manager[AcceptedKey.INTENT][column]
+        return intent
+
     def _prepare_transformation_pipeline(self, intent, column):
         # TODO this must be some optimization we can do. Walk through both
         #  intent override cases again to check how it affects the logic.
```
Comment (jichaogp) [inline foreshadow/steps/preprocessor.py]: These two functions look so similar...

File: foreshadow/utils/common.py [hunk 3]
```diff
@@ -199,3 +200,30 @@ def should_force_reresolve_based_on_override(self, X):
...
+    def __init__(self, column_name):
+        self.column_name = column_name
+
+    def fit(self, X, y=None):  # noqa
+        return self
+
+    def transform(self, X):  # noqa
+        if isinstance(X, DataFrame):
+            X["concat_col"] = X.apply(
+                lambda row: " ".join(row.dropna().tolist()), axis=1
```
Comment (jichaogp) [inline foreshadow/utils/common.py]: What is this trying to do?

File: foreshadow/smart/all.py [hunk 4]
```diff
@@ -382,38 +385,89 @@ def pick_transformer(self, X, y=None, **fit_params):
...
             strip_accents="unicode",
-            stop_words="english",
-            ngram_range=(1, 2),
-            max_df=0.9,
-            min_df=0.05,
-            max_features=None,
+            # stop_words="english",
+            # ngram_range=(1, 2),
+            # max_df=0.9,
+            min_df=0,
```
[the remark this answers is not in the document]
Comment (JingJZ160) [reply foreshadow/smart/all.py]: Oh these are the default arguments. I was testing them during a bug fix but they were irrelevant. I'll remove them.

File: foreshadow/steps/preprocessor.py (hunk 2 above)
Comment (JingJZ160) [reply foreshadow/steps/preprocessor.py]: I'll remove the second one. It was not used anymore.

File: foreshadow/steps/preprocessor.py (hunk 1 above)
Comment (JingJZ160) [reply foreshadow/steps/preprocessor.py]: Not here. There is a class called `DataSelector` where we merge text features. What happens here is that we send all text features to one transformation pipeline in one group.

File: foreshadow/utils/common.py (hunk 3 above)
Comment (JingJZ160) [reply foreshadow/utils/common.py]: Merge all the text columns with a space.

**Comment by JingJZ160** [comment]
The latest change includes a TruncatedSVD step in the text transformation pipeline. The tricky part is to decide how many components to use. As for now, the value is set to:

- The number of non-text features
- A default value of 20 (This is arbitrary. We can discuss further if needed).
- The number of features from TFIDF output minus 1. This is required by the SVD code. The number of components must be smaller than the number of features from the input, which is the sparse matrix output from TFIDF.

As for performance, I tested on the 20news_group data described here: https://scikit-learn.org/stable/tutorial/text_analytics/working_with_text_data.html Using Foreshadow running only 1 minute, it has a higher classification accuracy than the NB but lower than the SVM. If there is interest to do more test, I can let TPOT run longer. Let me know your thoughts.
